WalletWasabi/Tor/Socks5/Pool/TorHttpPool.cs

```diff
@@ -322,6 +321,29 @@ internal virtual async Task<HttpResponseMessage> SendCoreAsync(TorTcpConnection
 		}
 	}
 
+	/// <summary>Allows to report that a Tor circuit was closed.</summary>
+	/// <param name="circuitName">Name of the circuit. Example is: <c>IK1DG1HZCZFEQUTF86O86</c>.</param>
+	/// <remarks>
+	/// Useful to clean up <see cref="ConnectionPerHost"/> so that we do not exhaust <see cref="MaxConnectionsPerHost"/> limit.
+	/// <para>If client code forgets to dispose <see cref="PersonCircuit"/>, this should help us to recover eventually.</para>
+	/// </remarks>
+	public async Task ReportCircuitClosedAsync(string circuitName, CancellationToken cancellationToken)
+	{
+		using (await ObtainPoolConnectionLock.LockAsync(cancellationToken).ConfigureAwait(false))
+		{
+			foreach ((string host, List<TorTcpConnection> tcpConnections) in ConnectionPerHost)
+			{
+				foreach (TorTcpConnection tcpConnection in tcpConnections)
+				{
+					if (tcpConnection.Circuit.Name == circuitName)
```

Comment:
Shall we log if not found?

Reply:
Good question. Given that Tor has tor circuits with multiple purposes, we will most likely get non-trivial amount of "not found log lines" which does not seem useful.

We can add the log line but I would do it in a follow-up PR.